File: engine/helengine.core/assets/font/FontAssetBinarySerializer.cs
namespace helengine {
    /// <summary>
    /// Deserializes packaged font assets used by player builds.
    /// </summary>
    public static class FontAssetBinarySerializer {
        /// <summary>
        /// Shared format identifier for packaged font payloads.
        /// </summary>
        public const ushort FormatId = 1;

        /// <summary>
        /// Record kind used for serialized packaged font payloads.
        /// </summary>
        public const EditorBinaryRecordKind RecordKind = EditorBinaryRecordKind.FontAsset;

        /// <summary>
        /// Serializer version for the current packaged font payload layout.
        /// </summary>
        public const byte CurrentVersion = 5;

        /// <summary>
        /// Value kind used for serialized packaged font payloads.
        /// </summary>
        public const ushort ValueKind = 1;

        /// <summary>
        /// Payload description used when a stored header carries a foreign format id.
        /// </summary>
        const string FormatIdMismatchSubject = "font binary";

        /// <summary>
        /// Payload description used when a stored header carries an unexpected record kind or value kind.
        /// </summary>
        const string RecordMismatchSubject = "font";

        /// <summary>
        /// Payload description used when a stored header carries an unsupported serializer version.
        /// </summary>
        const string VersionMismatchSubject = "Packaged font";

        /// <summary>
        /// Recovery guidance appended to the version mismatch message for stale packaged fonts.
        /// </summary>
        const string RegenerateInstruction = "Regenerate the packaged font asset.";

        /// <summary>
        /// Gets the most recent font-deserialization stage reached by the packaged runtime loader.
        /// </summary>
        public static string LastDeserializeStage { get; private set; } = string.Empty;

        /// <summary>
        /// Deserializes a font asset from the supplied stream.
        /// </summary>
        /// <param name="stream">Source stream containing the packaged font.</param>
        /// <returns>Deserialized font asset.</returns>
        public static FontAsset Deserialize(Stream stream, RenderManager2D renderManager2D) {
            if (stream == null) {
                throw new ArgumentNullException(nameof(stream));
            }
            if (renderManager2D == null) {
                throw new ArgumentNullException(nameof(renderManager2D));
            }

            EngineBinaryHeader header = EngineBinaryHeaderSerializer.Read(stream);
            try {
                return Deserialize(stream, header, renderManager2D);
            } finally {
                NativeOwnership.Delete(header);
            }
        }

        /// <summary>
        /// Deserializes a font asset after the standardized header has already been read.
        /// </summary>
        /// <param name="stream">Source stream positioned at the payload.</param>
        /// <param name="header">Previously decoded HELE header.</param>
        /// <returns>Deserialized font asset.</returns>
        public static FontAsset Deserialize(Stream stream, [NativeNoEscape] EngineBinaryHeader header, RenderManager2D renderManager2D) {
            using EngineBinaryReader reader = VersionedBinaryPayload.ValidateHeader(
                stream,
                header,
                FormatId,
                (ushort)RecordKind,
                ValueKind,
                CurrentVersion,
                FormatIdMismatchSubject,
                RecordMismatchSubject,
                VersionMismatchSubject,
                VersionedBinaryVersionMismatchStyle.RequiredVersionSubjectFirst,
                RegenerateInstruction);

            if (renderManager2D == null) {
                throw new ArgumentNullException(nameof(renderManager2D));
            }

            string cookedAtlasTextureRelativePath;
            FontInfo fontInfo;
            float lineHeight;
            int atlasWidth;
            int atlasHeight;
            ulong sourceTextureRuntimeAssetId;
            ushort sourceTextureWidth;
            ushort sourceTextureHeight;
            TextureAssetColorFormat sourceTextureColorFormat;
            TextureAssetAlphaPrecision sourceTextureAlphaPrecision;
            byte[] sourceTexturePaletteColors;
            byte[] sourceTextureColors;

            cookedAtlasTextureRelativePath = reader.ReadString();
            sourceTextureRuntimeAssetId = (ulong)reader.ReadInt64();
            sourceTextureWidth = reader.ReadUInt16();
            sourceTextureHeight = reader.ReadUInt16();
            sourceTextureColorFormat = ReadTextureAssetColorFormat(reader);
            sourceTextureAlphaPrecision = ReadTextureAssetAlphaPrecision(reader);
            sourceTexturePaletteColors = reader.ReadByteArray();
            sourceTextureColors = reader.ReadByteArray();

            fontInfo = new FontInfo(
                reader.ReadString(),
                reader.ReadInt32(),
                reader.ReadSingle());

            lineHeight = reader.ReadSingle();
            atlasWidth = reader.ReadInt32();
            atlasHeight = reader.ReadInt32();

            TextureAsset sourceTexture = new TextureAsset {
                RuntimeAssetId = sourceTextureRuntimeAssetId,
                Width = sourceTextureWidth,
                Height = sourceTextureHeight,
                ColorFormat = sourceTextureColorFormat,
                AlphaPrecision = sourceTextureAlphaPrecision,
                PaletteColors = sourceTexturePaletteColors,
                Colors = sourceTextureColors
            };

            int characterCount = reader.ReadInt32();
            Dictionary<char, FontChar> characters = new Dictionary<char, FontChar>(characterCount);
            for (int index = 0; index < characterCount; index++) {
                char character = (char)reader.ReadUInt16();
                FontChar fontChar = new FontChar(
                    reader.ReadFloat4(),
                    reader.ReadSingle(),
                    reader.ReadSingle(),
                    reader.ReadSingle(),
                    reader.ReadSingle());
                characters.Add(character, fontChar);
            }

            if (sourceTexture.Width > 0 && sourceTexture.Height > 0 && sourceTexture.Colors != null && sourceTexture.Colors.Length > 0) {
                RuntimeTexture texture = renderManager2D.BuildTextureFromRaw(sourceTexture);
                return new FontAsset(fontInfo, texture, characters, lineHeight, atlasWidth, atlasHeight) {
                    SourceTextureAsset = sourceTexture,
                    CookedAtlasTextureRelativePath = cookedAtlasTextureRelativePath
                };
            }

            NativeOwnership.DisposeAndDelete(sourceTexture);
            return new FontAsset(fontInfo, null, characters, lineHeight, atlasWidth, atlasHeight) {
                CookedAtlasTextureRelativePath = cookedAtlasTextureRelativePath
            };
        }

        /// <summary>
        /// Reads one serialized texture color-format value from the packaged font payload.
        /// </summary>
        /// <param name="reader">Source reader positioned at the texture format byte.</param>
        /// <returns>Decoded texture color format.</returns>
        static TextureAssetColorFormat ReadTextureAssetColorFormat(EngineBinaryReader reader) {
            if (reader == null) {
                throw new ArgumentNullException(nameof(reader));
            }

            byte serializedValue = reader.ReadByte();
            if (serializedValue == (byte)TextureAssetColorFormat.Rgba32) {
                return TextureAssetColorFormat.Rgba32;
            } else if (serializedValue == (byte)TextureAssetColorFormat.Rgba4444) {
                return TextureAssetColorFormat.Rgba4444;
            } else if (serializedValue == (byte)TextureAssetColorFormat.Indexed4) {
                return TextureAssetColorFormat.Indexed4;
            } else if (serializedValue == (byte)TextureAssetColorFormat.Indexed8) {
                return TextureAssetColorFormat.Indexed8;
            } else if (serializedValue == (byte)TextureAssetColorFormat.GxRgb5A3) {
                return TextureAssetColorFormat.GxRgb5A3;
            }

            throw new InvalidOperationException($"Unsupported texture color format '{serializedValue}'.");
        }

        /// <summary>
        /// Reads one serialized texture alpha-precision value from the packaged font payload.
        /// </summary>
        /// <param name="reader">Source reader positioned at the texture alpha-precision byte.</param>
        /// <returns>Decoded texture alpha precision.</returns>
        static TextureAssetAlphaPrecision ReadTextureAssetAlphaPrecision(EngineBinaryReader reader) {
            if (reader == null) {
                throw new ArgumentNullException(nameof(reader));
            }

            byte serializedValue = reader.ReadByte();
            if (serializedValue == (byte)TextureAssetAlphaPrecision.Opaque) {
                return TextureAssetAlphaPrecision.Opaque;
            } else if (serializedValue == (byte)TextureAssetAlphaPrecision.Binary) {
                return TextureAssetAlphaPrecision.Binary;
            } else if (serializedValue == (byte)TextureAssetAlphaPrecision.A4) {
                return TextureAssetAlphaPrecision.A4;
            } else if (serializedValue == (byte)TextureAssetAlphaPrecision.A8) {
                return TextureAssetAlphaPrecision.A8;
            }

            throw new InvalidOperationException($"Unsupported texture alpha precision '{serializedValue}'.");
        }

    }
}

File: tmp/helengine-core-cpp-regenerated/FontAssetBinarySerializer.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "FontAssetBinarySerializer.hpp"
#include "runtime/native_exceptions.hpp"
#include "EngineBinaryHeader.hpp"
#include "EngineBinaryHeaderSerializer.hpp"
#include "FontAsset.hpp"
#include "EngineBinaryReader.hpp"
#include "FontInfo.hpp"
#include "TextureAsset.hpp"
#include "runtime/native_dictionary.hpp"
#include "RuntimeTexture.hpp"
#include "Core.hpp"
#include "RenderManager2D.hpp"
#include "EditorBinaryRecordKind.hpp"
#include "FontChar.hpp"
#include "runtime/array.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_dictionary.hpp"
#include "runtime/native_disposable.hpp"
#include "runtime/native_enum.hpp"
#include "runtime/native_event.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_nullable.hpp"
#include "runtime/native_span.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_tuple.hpp"
#include "runtime/native_type.hpp"
#include "system/app_context.hpp"
#include "system/bit_converter.hpp"
#include "system/guid.hpp"
#include "system/io/file-stream.hpp"
#include "system/io/file.hpp"
#include "system/io/memory-stream.hpp"
#include "system/io/path.hpp"
#include "system/io/stream.hpp"
#include "system/string_comparer.hpp"
#include "system/text/encoding.hpp"

uint8_t FontAssetBinarySerializer::CurrentVersion = 1;

uint16_t FontAssetBinarySerializer::FormatId = 1;

::EditorBinaryRecordKind FontAssetBinarySerializer::RecordKind = EditorBinaryRecordKind::FontAsset;

::FontAsset* FontAssetBinarySerializer::Deserialize(::Stream* stream)
{
    if (stream == nullptr)
    {
throw new ArgumentNullException("stream");
    }
::EngineBinaryHeader *header = EngineBinaryHeaderSerializer::Read(stream);
return Deserialize(stream, header);}

::FontAsset* FontAssetBinarySerializer::Deserialize(::Stream* stream, ::EngineBinaryHeader* header)
{
    if (stream == nullptr)
    {
throw new ArgumentNullException("stream");
    }
    if (header == nullptr)
    {
throw new ArgumentNullException("header");
    }
ValidateHeader(header);
    if (Core::get_Instance() == nullptr || Core::get_Instance()->get_RenderManager2D() == nullptr)
    {
throw new InvalidOperationException("Font assets require an initialized core renderer before deserialization.");
    }
{
::EngineBinaryReader *reader = EngineBinaryReader::Create(stream, header->get_Endianness(), true);
::FontInfo *fontInfo = ([&]() {
auto __ctor_arg_00000059 = reader->ReadString();
auto __ctor_arg_0000005A = reader->ReadInt32();
auto __ctor_arg_0000005B = reader->ReadSingle();
return new ::FontInfo(__ctor_arg_00000059, __ctor_arg_0000005A, __ctor_arg_0000005B);
})();
const float lineHeight = reader->ReadSingle();
const int32_t atlasWidth = reader->ReadInt32();
const int32_t atlasHeight = reader->ReadInt32();
::TextureAsset *sourceTexture = ([&]() {
auto __object_0000005C = new ::TextureAsset();
__object_0000005C->Width = reader->ReadUInt16();
__object_0000005C->Height = reader->ReadUInt16();
__object_0000005C->Colors = reader->ReadByteArray();
return __object_0000005C;
})();
const int32_t characterCount = reader->ReadInt32();
Dictionary<char, ::FontChar> *characters = new Dictionary<char, ::FontChar>(characterCount);
for (int32_t index = 0; index < characterCount; index++) {
const char character = static_cast<char>(reader->ReadUInt16());
::FontChar fontChar = ([&]() {
auto __ctor_arg_0000005D = reader->ReadFloat4();
auto __ctor_arg_0000005E = reader->ReadSingle();
auto __ctor_arg_0000005F = reader->ReadSingle();
auto __ctor_arg_00000060 = reader->ReadSingle();
auto __ctor_arg_00000061 = reader->ReadSingle();
return ::FontChar(__ctor_arg_0000005D, __ctor_arg_0000005E, __ctor_arg_0000005F, __ctor_arg_00000060, __ctor_arg_00000061);
})();
characters->Add(character, fontChar);
}
::RuntimeTexture *texture = Core::get_Instance()->get_RenderManager2D()->BuildTextureFromRaw(sourceTexture);
::FontAsset *asset = ([&]() {
auto __object_00000062 = new ::FontAsset(fontInfo, texture, characters, lineHeight, atlasWidth, atlasHeight);
__object_00000062->set_SourceTextureAsset(sourceTexture);
return __object_00000062;
})();
return asset;}
}

void FontAssetBinarySerializer::ValidateHeader(::EngineBinaryHeader* header)
{
    if (header->get_FormatId() != FormatId)
    {
throw new InvalidOperationException(std::string("Unsupported font binary format id '") + std::to_string(header->get_FormatId()) + std::string("'."));
    }
    if (header->get_RecordKind() != static_cast<uint16_t>(RecordKind))
    {
throw new InvalidOperationException(std::string("Unexpected font record kind '") + std::to_string(header->get_RecordKind()) + std::string("'."));
    }
    if (header->get_Version() != CurrentVersion)
    {
throw new InvalidOperationException(std::string("Unsupported font binary version '") + std::to_string(header->get_Version()) + std::string("'."));
    }
}

